ReverseTicTacToe/UI/ConsoleUI.cs

```diff
@@ -33,17 +29,145 @@ public  void Init()
+            string turnMsg = GetTurnMessage();
+            
+            if(m_Game.CurrentTurn == 1 || m_Game.GameType == Game.eGameTypes.PVP)
+            {
+                string userTurnInput;
+                do
+                {
+                    WriteLine(turnMsg);
+                    userTurnInput = ReadLine();
+                }
+                while(!turnInputIsValid(userTurnInput, out row, out col));
+            }
+            if (row == -1 && col == -1)
+            {
+                //quit game
+            }
+            m_Game.MakeTurn(row, col);
+        }
+
+        private bool turnInputIsValid(string i_UserTurnInput, out int i_Row, out int i_Col)
+        {
+            i_Row = -1;
+            i_Col = -1;
+            const bool v_IsValid = true;
+            bool turnInputIsValid = !v_IsValid;
+            char[] charsToTrim = { ' ', '\t' };
+            i_UserTurnInput = i_UserTurnInput.Trim(charsToTrim);
+            if(i_UserTurnInput == "Q")
+            {
+                turnInputIsValid = v_IsValid;
+            }
+            else if (i_UserTurnInput.Length == 3 && i_UserTurnInput[1] == ',')
+            {
+                char[] charArray = i_UserTurnInput.ToCharArray();
+                if(char.IsDigit(charArray[0]) && char.IsDigit(charArray[2]) && charArray[1] == ',')
+                {
+                    int digit1 = int.Parse(charArray[0].ToString());
+                    int digit2 = int.Parse(charArray[2].ToString());
+                    if(isInRage(digit1,0,m_Game.BoardSize) &&
+                        (isInRage(digit2,0,m_Game.BoardSize)))
+                    {
+                        turnInputIsValid = v_IsValid;
+                        i_Row = digit1;
+                        i_Col = digit2;
+                    }
+                }
+            }
+            return turnInputIsValid;
+        }
+
+        private bool isInRage(int i_Num, int i_Min, int i_Max)
+        {
+            return i_Num >= i_Min && i_Num <= i_Max;
+        }
+
+        public string GetTurnMessage()
+        {
+            string TurnMsg;
+            if((Game.eTurns)m_Game.CurrentTurn == Game.eTurns.PLAYER1)
+            {
+                TurnMsg = "Player 1 your turn, please enter where to place your symbol (row, col): "
+                          + Environment.NewLine;
+            }
+            else
+            {
+                if(m_Game.GameType == Game.eGameTypes.PVP)
+                {
+                    TurnMsg = "Player 2 your turn, please enter where to place your symbol (row, col): "
+                              + Environment.NewLine;
+                }
+                else
+                {
+                    TurnMsg = "The computer is now calculating what's the best next move is:"
+                              + Environment.NewLine;
+                }
+            }
+            return TurnMsg;
+        }
+
+        private void printGameBoard()
```

Comment:
I think maybe that method should be a service provided by the logic layer, not the UI. the CosoleUI should just call that method, which will return a string represantation of the game board, and the consoleUI will print it . what do you think?

Reply:
Let's say that we will need the next version to make it with GUI and WPF graphics elements. 
then this function that will be in the logic part won't be used.
I mean - this function that creates a string representation of the board is only relevant for Console GUI version and that's in the UI department IMO

Reply:
GOOOOOOD POINT